### PR TITLE
Fix line breakpoints for `return` statements without a value

### LINKED ISSUE #17081: Breakpoint set on return doesn't work properly
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Putting breakpoint on return will stop on next line rather on return itself. 

```powershell
function TestMe {
    [cmdletBinding()]
    param(

    )

    $Test = Get-Verb

    Write-Host "Test"
    return
}

function TestMe2 {
    [cmdletBinding()]
    param(

    )

    $Test = Get-Verb


    if ($test.count -eq 5) {
        Write-Host "test1"
        return
    } elseif ($test.count -eq 1) {
        Write-Host "test"
        return
    }
    Write-Host "Test"
    return
}

TestMe2
```

image

Originally reported in https://github.com/PowerShell/vscode-powershell/issues/3896

According to @SeeminglyScience it's a PowerShell bug:

image

This gets even worse if you try to do a blank line after return and put a breakpoint on return. 

```powershell
function TestMe2 {
    [cmdletBinding()]
    param(

    )

    $Test = Get-Verb


    if ($test.count -eq 5) {
        Write-Host "test1"
        return
    } elseif ($test.count -eq 1) {
        Write-Host "test"
        return
    }
    Write-Host "Test"
    return

}

TestMe2
```

Breakpoint won't be hit. If you put `return ""` breakpoint will work properly. 

### Expected behavior

```console
Breakpoint should be at return as expected.
```

### Actual behavior

```console
Breakpoint stops 1 line after expected place
```

### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.0-preview.3
PSEdition                      Core
GitCommitId                    7.3.0-preview.3
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

But is there also for PS 5.1
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by SeeminglyScience**
Fixes #17081

# PR Summary

Adds an explicit call to `UpdatePosition` when a `return` statement does not include a pipeline.

Codegen before change:

```csharp
// Generated with:
// { return } | Get-ScriptBlockDisassembly -IgnoreStartupAndTeardown
// ScriptBlock.EndBlock
funcContext._currentSequencePointIndex = 0;
try
{
    return ;
}
catch (FlowControlException)
{
    throw;
}
catch (Exception exception)
{
    ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
}
funcContext._currentSequencePointIndex = 1;

if (context._debuggingMode > 0)
{
    context._debugger.OnSequencePointHit(funcContext);
}
```

Codegen after change:

```csharp
// Generated with:
// { return } | Get-ScriptBlockDisassembly -IgnoreStartupAndTeardown
// ScriptBlock.EndBlock
funcContext._currentSequencePointIndex = 0;
try
{
    funcContext._currentSequencePointIndex = 1;

    if (context._debuggingMode > 0)
    {
        context._debugger.OnSequencePointHit(funcContext);
    }
    return ;
}
catch (FlowControlException)
{
    throw;
}
catch (Exception exception)
{
    ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
}
funcContext._currentSequencePointIndex = 2;

if (context._debuggingMode > 0)
{
    context._debugger.OnSequencePointHit(funcContext);
}
```

<details>
<summary>

## Trap compilation (click to expand)</summary>

Turns out there wasn't an easy way to view disassembly of a trap with `ScriptBlockDisassembler`. So that took a minute to figure out how to do easily:

<details>
<summary>

### Code to generate (click to expand)</summary>

```powershell
#requires -Modules ImpliedReflection, ScriptBlockDisassembler

Enable-ImpliedReflection -YesIKnowIShouldNotDoThis

$sb = {
    trap { return }
    throw
}

$trap = $sb.Ast.EndBlock.Traps[0]
$main = [System.Management.Automation.Language.Compiler]::new()
$main.Compile($sb.ScriptBlockData, $true)
$secondary = [System.Management.Automation.Language.Compiler]::new(
    $main._sequencePoints,
    $main._sequencePointIndexMap)

$secondary._compilingTrap = $true
$funcName = $main._currentFunctionName + '<trap>'
if ($null -ne $trap.TrapType) {
    $funcName += '<' + $trap.TrapType.TypeName.Name + '>'
}

$tuple = [Management.Automation.Language.VariableAnalysis]::AnalyzeTrap($trap)
$secondary.LocalVariablesTupleType = $tuple.Item1
$secondary.LocalVariablesParameter = [Linq.Expressions.Expression]::Variable(
    $secondary.LocalVariablesTupleType,
    'locals')

$lambda = $secondary.CompileSingleLambda(
    $trap.Body.Statements,
    $trap.Body.Traps,
    $funcName,
    <# entryExtent: #> $null,
    <# exitExtent: #> $null,
    <# rootForDefiningTypesAndUsings: #> $null)

$lambda | Format-ExpressionTree | bat -l cs
```
</details>

### Codegen before:

```csharp
(FunctionContext funcContext) =>
{
    try
    {
        ExecutionContext context = funcContext._executionContext;
        MutableTuple<object, object[], object, object, PSScriptCmdlet, PSBoundParametersDictionary, InvocationInfo, string, string, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null> locals = (MutableTuple<object, object[], object, object, PSScriptCmdlet, PSBoundParametersDictionary, InvocationInfo, string, string, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null>)funcContext._localsTuple;
        funcContext._functionName = "<ScriptBlock><trap>";
        context._debugger.EnterScriptFunction(funcContext);
        try
        {
            throw new ReturnException(AutomationNull.Value);
        }
        catch (FlowControlException)
        {
            throw;
        }
        catch (Exception exception)
        {
            ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
        }
    }
    finally
    {
        context._debugger.ExitScriptFunction();
    }
}
```

### Codegen after:

```csharp
(FunctionContext funcContext) =>
{
    try
    {
        ExecutionContext context = funcContext._executionContext;
        MutableTuple<object, object[], object, object, PSScriptCmdlet, PSBoundParametersDictionary, InvocationInfo, string, string, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null> locals = (MutableTuple<object, object[], object, object, PSScriptCmdlet, PSBoundParametersDictionary, InvocationInfo, string, string, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null, LanguagePrimitives.Null>)funcContext._localsTuple;
        funcContext._functionName = "<ScriptBlock><trap>";
        context._debugger.EnterScriptFunction(funcContext);
        try
        {
            funcContext._currentSequencePointIndex = 1;

            if (context._debuggingMode > 0)
            {
                context._debugger.OnSequencePointHit(funcContext);
            }
            throw new ReturnException(AutomationNull.Value);
        }
        catch (FlowControlException)
        {
            throw;
        }
        catch (Exception exception)
        {
            ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
        }
    }
    finally
    {
        context._debugger.ExitScriptFunction();
    }
}
```

</details>

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1 [hunk 1]
```diff
@@ -331,6 +331,38 @@ elseif (Test-Path $PSCommandPath)
...
     }
+
+    Context "Break point on return should hit" {
+        BeforeAll {
+            $return_script_1 = @'
+return
+'@
+            $return_script_2 = @'
+return 10
+'@
```
Comment (daxian-dbw) [inline test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1:342]: Can you please add one more test case to cover `return` statement in a trap block?

**Review by daxian-dbw — COMMENTED** [review]
LGTM with a comment

File: test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1 (hunk 1 above)
Comment (SeeminglyScience) [reply test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1:342]: Done! Also added disassembly for traps in the description.

Comment (SeeminglyScience) [reply test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1:342]: Probably should have waited for the tests to pass before saying that. Looking into the failure

Comment (daxian-dbw) [reply test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1:342]: Thanks for the additional fix. Will take a look today.

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**